crates/polars-stream/src/nodes/parquet_source.rs

```diff
@@ -0,0 +1,1734 @@
+};
+use polars_core::schema::IndexOfSchema;
+use polars_core::series::{IntoSeries, IsSorted, Series};
+use polars_core::utils::operation_exceeded_idxsize_msg;
+use polars_error::{polars_bail, polars_err, PolarsResult};
+use polars_expr::prelude::PhysicalExpr;
+use polars_io::cloud::CloudOptions;
+use polars_io::predicates::PhysicalIoExpr;
+use polars_io::prelude::{FileMetaData, ParquetOptions};
+use polars_io::utils::byte_source::{
+    ByteSource, DiscreteRangesByteSource, DynByteSource, DynByteSourceBuilder, MemSliceByteSource,
+};
+use polars_io::utils::slice::SplitSlicePosition;
+use polars_io::{is_cloud_url, RowIndex};
+use polars_parquet::read::RowGroupMetaData;
+use polars_plan::plans::hive::HivePartitions;
+use polars_plan::plans::FileInfo;
+use polars_plan::prelude::FileScanOptions;
+use polars_utils::mmap::MemSlice;
+use polars_utils::slice::GetSaferUnchecked;
+use polars_utils::IdxSize;
+
+use super::{MorselSeq, TaskPriority};
+use crate::async_executor::{self};
+use crate::async_primitives::distributor_channel::{self, distributor_channel};
+use crate::morsel::get_ideal_morsel_size;
+use crate::utils::notify_channel::{notify_channel, NotifyReceiver};
+use crate::utils::task_handles_ext;
+
+#[allow(clippy::type_complexity)]
+pub struct ParquetSourceNode {
+    paths: Arc<Vec<PathBuf>>,
+    file_info: FileInfo,
+    hive_parts: Option<Arc<Vec<HivePartitions>>>,
+    predicate: Option<Arc<dyn PhysicalExpr>>,
+    options: ParquetOptions,
+    cloud_options: Option<CloudOptions>,
+    file_options: FileScanOptions,
+    // Run-time vars
+    config: Config,
+    verbose: bool,
+    physical_predicate: Option<Arc<dyn PhysicalIoExpr>>,
+    projected_arrow_fields: Arc<[polars_core::prelude::ArrowField]>,
+    byte_source_builder: DynByteSourceBuilder,
+    memory_prefetch_func: fn(&[u8]) -> (),
+    // This permit blocks execution until the first morsel is requested.
+    row_group_data_task_starter: Option<tokio::sync::oneshot::Sender<()>>,
+    morsel_receivers: Vec<distributor_channel::Receiver<(DataFrame, MorselSeq)>>,
+    row_group_data_task_handle: Option<async_executor::AbortOnDropHandle<PolarsResult<()>>>,
+    row_group_decoder: Option<Arc<RowGroupDecoder>>,
+    is_finished: Arc<AtomicBool>,
+}
+
+#[allow(clippy::too_many_arguments)]
+impl ParquetSourceNode {
+    pub fn new(
+        paths: Arc<Vec<PathBuf>>,
+        file_info: FileInfo,
+        hive_parts: Option<Arc<Vec<HivePartitions>>>,
+        predicate: Option<Arc<dyn PhysicalExpr>>,
+        options: ParquetOptions,
+        cloud_options: Option<CloudOptions>,
+        file_options: FileScanOptions,
+    ) -> Self {
+        let verbose = config::verbose();
+
+        let byte_source_builder =
+            if is_cloud_url(paths[0].to_str().unwrap()) || config::force_async() {
+                DynByteSourceBuilder::ObjectStore
+            } else {
+                DynByteSourceBuilder::Mmap
```

Comment:
Mmap is currently default, let me know if we want to change

Reply:
Let's investigate that in a future PR.